turbo/rpchelper/filters.go

```diff
@@ -576,97 +511,80 @@ func (ff *Filters) OnNewTx(reply *txpool.OnAddReply) {
 			break
 		}
 	}
-	for _, v := range ff.pendingTxsSubs {
-		v <- txs
-	}
+	ff.pendingTxsSubs.Range(func(k PendingTxsSubID, v Sub[[]types.Transaction]) error {
+		v.Send(txs)
+		return nil
+	})
 }
 
 func (ff *Filters) OnNewLogs(reply *remote.SubscribeLogsReply) {
 	ff.logsSubs.distributeLog(reply)
 }
 
-func generateSubscriptionID() SubscriptionID {
-	var id [32]byte
-
-	_, err := rand.Read(id[:])
-	if err != nil {
-		log.Crit("rpc filters: error creating random id", "err", err)
-	}
-
-	return SubscriptionID(fmt.Sprintf("%x", id))
-}
-
 func (ff *Filters) AddLogs(id LogsSubID, logs *types.Log) {
-	ff.storeMu.Lock()
-	defer ff.storeMu.Unlock()
-	st, ok := ff.logsStores[id]
-	if !ok {
-		st = make([]*types.Log, 0)
-	}
-	st = append(st, logs)
-	ff.logsStores[id] = st
+	ff.logsStores.DoAndStore(id, func(st []*types.Log, ok bool) []*types.Log {
+		if !ok {
+			st = make([]*types.Log, 0)
+		}
+		st = append(st, logs)
+		return st
+	})
 }
 
 func (ff *Filters) ReadLogs(id LogsSubID) ([]*types.Log, bool) {
-	ff.storeMu.Lock()
-	defer ff.storeMu.Unlock()
 	res := make([]*types.Log, 0)
-	st, ok := ff.logsStores[id]
-	if !ok {
-		return res, false
-	}
-	res = append(res, st...)
-	st = make([]*types.Log, 0)
-	ff.logsStores[id] = st
-	return res, true
+	_, exist := ff.logsStores.Do(id, func(l []*types.Log, ok bool) ([]*types.Log, bool) {
+		if !ok {
+			return res, false
+		}
+		res = append(res, l...)
+		return []*types.Log{}, true
+	})
+	return res, exist
 }
 
 func (ff *Filters) AddPendingBlock(id HeadsSubID, block *types.Header) {
-	ff.storeMu.Lock()
-	defer ff.storeMu.Unlock()
-	st, ok := ff.pendingHeadsStores[id]
-	if !ok {
-		st = make([]*types.Header, 0)
-	}
-	st = append(st, block)
-	ff.pendingHeadsStores[id] = st
+	ff.pendingHeadsStores.DoAndStore(id, func(st []*types.Header, ok bool) []*types.Header {
+		if !ok {
+			st = make([]*types.Header, 0)
+		}
+		st = append(st, block)
+		return st
+	})
 }
 
 func (ff *Filters) ReadPendingBlocks(id HeadsSubID) ([]*types.Header, bool) {
-	ff.storeMu.Lock()
-	defer ff.storeMu.Unlock()
 	res := make([]*types.Header, 0)
-	st, ok := ff.pendingHeadsStores[id]
-	if !ok {
-		return res, false
-	}
-	res = append(res, st...)
-	st = make([]*types.Header, 0)
-	ff.pendingHeadsStores[id] = st
-	return res, true
+	_, exist := ff.pendingHeadsStores.Do(id, func(st []*types.Header, ok bool) ([]*types.Header, bool) {
+		if !ok {
+			return res, false
+		}
+		res = append(res, st...)
+		st = make([]*types.Header, 0)
```

Comment:
don't understand this line

Reply:
this was in the old function

I think it is attempting to make a copy of res to return, and then clear?

i will make it more explicit what it is doing.

Reply:
i've updated these to just call delete, since that's effectively what they do.